hooks/hookexecution/execution.go

```diff
@@ -311,3 +315,40 @@ func handleHookMutations[P any](
 
 	return payload
 }
+
+func handleModuleActivities[T any, P any](hook hooks.HookWrapper[T], activityControl privacy.ActivityControl, payload P) P {
+	// only 2 stages receive bidder request: hookstage.ProcessedAuctionRequestPayload and hookstage.BidderRequestPayload
+	// they both implement PayloadBidderRequest interface in order to execute mutations on bid request
+	var payloadData hookstage.PayloadBidderRequest
+	ok := true
+	if payloadData, ok = any(&payload).(hookstage.PayloadBidderRequest); !ok {
+		// payload doesn't have a bid request
+		return payload
+	}
+
+	privacyEnforcement := privacy.Enforcement{}
+
+	// parse hook.Module to split it to type and mame?
+	// hook.Module example: "mytest.mymodule". Can it be "rtd.mymodule" or "general.mymodule"?
+	scopeGeneral := privacy.Component{Type: privacy.ComponentTypeGeneral, Name: hook.Code} ///!!!!! hook.Code?
+	transmitUserFPDActivityAllowed := activityControl.Allow(privacy.ActivityTransmitUserFPD, scopeGeneral)
+	if !transmitUserFPDActivityAllowed {
+		privacyEnforcement.UFPD = true
+	}
+
+	if privacyEnforcement.AnyActivities() {
+		// changes need to be applied to new payload and leave original payload unchanged
+		bidderReq := payloadData.GetBidderRequestPayload()
+		var bidderReqCopy *openrtb2.BidRequest
+		bidderReqCopy = ptrutil.Clone(bidderReq)
+
+		privacyEnforcement.Apply(bidderReqCopy)
+
+		var newPayload = payload
+		var np = any(&newPayload).(hookstage.PayloadBidderRequest)
+		np.SetBidderRequestPayload(bidderReqCopy)
+		return newPayload
```

Comment:
What's the point of newPayload? Does it make a shallow copy?

Reply:
At this point we know already that request will be changed for this module (for instance user.ID will be removed). 
We create a shallow copy before "Apply" and inside of the "Apply" new User will be created and set back to request so  this module will not receive restricted data. We don't do this if activity will not change the request.

```go
      if privacyEnforcement.AnyActivities() {
		// changes need to be applied to new payload and leave original payload unchanged
		bidderReq := payloadData.GetBidderRequestPayload()
		bidderReqCopy := ptrutil.Clone(bidderReq)

		privacyEnforcement.Apply(bidderReqCopy)

		var newPayload = payload
		var np = any(&newPayload).(hookstage.PayloadBidderRequest)
		np.SetBidderRequestPayload(bidderReqCopy)
		return newPayload
	}
```

Reply:
So you have a shallow copy of the bidrequest. That shallow copy includes a pointer to the original `User` object. If you remove the `ID` from the `User` object, you will be modifying the original request.

Reply:
What you are saying is correct. In this case User copy will be handled inside of Apply function.  It will execute `ScrubRequest` function and it will make copy of User at the beginning and then set it back to request 
```go
var userCopy *openrtb2.User
userCopy = ptrutil.Clone(bidRequest.User)
...
...
bidRequest.User = userCopy
return bidRequest
```